src/com/owncloud/android/files/services/FileUploadService.java

```diff
@@ -0,0 +1,1359 @@
+     */
+    public static final String KEY_CREATE_REMOTE_FOLDER = "CREATE_REMOTE_FOLDER";
+    /**
+     * Set to true if upload is to performed only when connected via wifi.
+     */
+    public static final String KEY_WIFI_ONLY = "WIFI_ONLY";
+    /**
+     * Set to true if upload is to performed only when phone is being charged.
+     */
+    public static final String KEY_WHILE_CHARGING_ONLY = "KEY_WHILE_CHARGING_ONLY";
+    /**
+     * Set to future UNIX timestamp. Upload will not be performed before this timestamp. 
+     */
+    public static final String KEY_UPLOAD_TIMESTAMP= "KEY_UPLOAD_TIMESTAMP";
+    
+    public static final String KEY_LOCAL_BEHAVIOUR = "BEHAVIOUR";
+
+    /**
+     * Describes local behavior for upload.
+     */
+    public enum LocalBehaviour {
+        /**
+         * Creates a copy of file and stores it in tmp folder inside owncloud
+         * folder on sd-card. After upload it is moved to local owncloud
+         * storage. Original file stays untouched.
+         */
+        LOCAL_BEHAVIOUR_COPY(0),
+        /**
+         * Upload file from current storage. Afterwards original file is move to
+         * local owncloud storage.
+         */
+        LOCAL_BEHAVIOUR_MOVE(1),
+        /**
+         * Just uploads file and leaves it where it is. Original file stays
+         * untouched.
+         */
+        LOCAL_BEHAVIOUR_FORGET(2);
+        private final int value;
+
+        private LocalBehaviour(int value) {
+            this.value = value;
+        }
+
+        public int getValue() {
+            return value;
+        }
+    }
+    
+    public enum UploadSingleMulti {
+        UPLOAD_SINGLE_FILE(0), UPLOAD_MULTIPLE_FILES(1);
+        private final int value;
+
+        private UploadSingleMulti(int value) {
+            this.value = value;
+        }
+
+        public int getValue() {
+            return value;
+        }
+    };
+
+    private static final String TAG = FileUploadService.class.getSimpleName();
+
+    private IBinder mBinder;
+    private OwnCloudClient mUploadClient = null;
+    private Account mLastAccount = null;
+    private FileDataStorageManager mStorageManager;
+    //since there can be only one instance of an Android service, there also just one db connection.
+    private UploadDbHandler mDb = null;
+    
+    private final AtomicBoolean mCancellationRequested = new AtomicBoolean(false);
+    private final AtomicBoolean mCancellationPossible = new AtomicBoolean(false);
+
+    /**
+     * List of uploads that are currently pending. Maps from remotePath to where file
+     * is being uploaded to {@link UploadFileOperation}.
+     */
+    private ConcurrentMap<String, UploadDbObject> mPendingUploads = new ConcurrentHashMap<String, UploadDbObject>();
+    
+    /**
+     * {@link UploadFileOperation} object of ongoing upload. Can be null. Note: There can only be one concurrent upload!
+     */
+    private UploadFileOperation mCurrentUpload = null;
```

Comment:
I am extremely glad to see you recovered ```UploadFileOperation``` instead of using an anonymous ```Runnable```, as I remember from past versions of this PR. I'm not sure if it was just due to necessity of merging develop, but I really appreciate it's back.

Anonymous classes are fine for writing some fast experiments, but they are not reusable at all. Operation classes may be run in different contexts (```Service```s, ```AsyncTask```s, plain brackground ```Thread```s) without copying and pasting code, what allows play with partial changes in the architecture, or just increase reusability through the app.